### PR TITLE
feat: add auth support via command

### PR DESCRIPTION
**Pull request by radulucut**
Implementation for https://github.com/jsdelivr/globalping-cli/issues/108

### CODE REVIEW

File: globalping/auth.go [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+	go func() {
+		ports := []int{60000, 60010, 60020, 60030, 60040, 60100, 60110, 60120, 60130, 60140}
+		var err error
+		for _, port := range ports {
+			server.Addr = fmt.Sprintf("localhost:%d", port)
+			err = server.ListenAndServe()
+			if err == nil {
+				break
+			}
+		}
```
Comment (MartinKolarik) [inline globalping/auth.go:82]: Not sure if it's an issue with this or with tests, but on Windows (also on Linux via WSL) the tests are currently failing for me with:

```
?       github.com/jsdelivr/globalping-cli      [no test files]
?       github.com/jsdelivr/globalping-cli/globalping/probe     [no test files]
?       github.com/jsdelivr/globalping-cli/mocks        [no test files]
?       github.com/jsdelivr/globalping-cli/packagecloud [no test files]
?       github.com/jsdelivr/globalping-cli/utils        [no test files]
?       github.com/jsdelivr/globalping-cli/version      [no test files]
ok      github.com/jsdelivr/globalping-cli/cmd  (cached)
--- FAIL: Test_Authorize (0.00s)
    auth_test.go:78: Post "http://localhost:60000/callback?code=cod3": dial tcp 127.0.0.1:60000: connect: connection refused
FAIL
FAIL    github.com/jsdelivr/globalping-cli/globalping   0.277s
ok      github.com/jsdelivr/globalping-cli/storage      (cached)
ok      github.com/jsdelivr/globalping-cli/view (cached)
FAIL
```

File: cmd/auth.go [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+		Short: "Authenticate with the Globalping API",
+		Long:  "Authenticate with the Globalping API for higher measurements limits.",
+	}
+
+	loginCmd := &cobra.Command{
+		RunE:  r.RunAuthLogin,
+		Use:   "login",
+		Short: "Log in to your Globalping account",
+		Long:  `Log in to your Globalping account for higher measurements limits.`,
+	}
```
Comment (MartinKolarik) [inline cmd/auth.go:23]: If the user is already logged in, silently call logout in the background to revoke the previous token.

File: cmd/auth.go [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+func (r *Root) RunAuthStatus(cmd *cobra.Command, args []string) error {
+	res, err := r.client.TokenIntrospection("")
+	if err != nil {
+		return err
+	}
+	if res.Active {
+		r.printer.Printf("Logged in as %s.\n", res.Username)
+	} else {
+		r.printer.Println("Not logged in.")
+	}
```
Comment (MartinKolarik) [inline cmd/auth.go:80]: We should get here if there's no token stored but currently introspection above fails and we exit at line 74.

File: cmd/auth.go [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+		defer func() {
+			r.cancel <- syscall.SIGINT
+		}()
+		if e != nil {
+			err = e
+			return
+		}
+		r.printer.Println("You are now authenticated")
+	})
+	r.printer.Println("Please visit the following URL to authenticate:")
```
Comment (MartinKolarik) [inline cmd/auth.go:65]: Can we also automatically open this in the default browser? (ignoring any errors if that fails)

**Review by MartinKolarik — COMMENTED** [review]
I think this part is missing?

> when the access token seems valid, but you got a 401 or 403 error from the API, and the refresh token is still valid, try refreshing the tokens
if this fails as well, delete the stored tokens
if this works, print a message for the user like "Access token successfully refreshed. Try repeating the measurement."

File: cmd/auth.go (hunk 3 above)
Comment (radulucut) [reply cmd/auth.go:80]: Not quite sure what you mean here?

**Comment by radulucut** [comment]
> I think this part is missing?
> 
> > when the access token seems valid, but you got a 401 or 403 error from the API, and the refresh token is still valid, try refreshing the tokens
> > if this fails as well, delete the stored tokens
> > if this works, print a message for the user like "Access token successfully refreshed. Try repeating the measurement."

The access token is silently refreshed if there's a refresh token. I didn't add any message because it might broke the live view, but I can add it if you think is worth it?

File: cmd/auth.go (hunk 3 above)
Comment (MartinKolarik) [reply cmd/auth.go:80]: Try running `auth status` with no token. It should print this message but right now prints an error instead.

**Comment by MartinKolarik** [comment]
> The access token is silently refreshed if there's a refresh token. I didn't add any message because it might broke the live view, but I can add it if you think is worth it?

Is it actually refreshed *after* the measurement fails with an error? I might have missed it, but I haven't seen that logic in the code. I understand the token is refreshed correctly when it expires, but this is about a case when it stops working before the original expiration date.

**Comment by radulucut** [comment]
> > The access token is silently refreshed if there's a refresh token. I didn't add any message because it might broke the live view, but I can add it if you think is worth it?
> 
> Is it actually refreshed _after_ the measurement fails with an error? I might have missed it, but I haven't seen that logic in the code. I understand the token is refreshed correctly when it expires, but this is about a case when it stops working before the original expiration date.

yeah, you're right, I've missed that. Thanks

File: globalping/auth.go (hunk 1 above)
Comment (radulucut) [reply globalping/auth.go]: It should work now

File: utils/utils.go [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+}
+
+func (u *utils) Now() _time.Time {
+	return _time.Now()
+}
+
+func (u *utils) OpenBrowser(url string) error {
+	switch runtime.GOOS {
+	case "linux":
+		return exec.Command("xdg-open", url).Start()
```
Comment (MartinKolarik) [inline utils/utils.go:29]: Can we additionally add WSL detection here? It's reported as Linux, but the command to open should like on Windows (but needs the `exe` extension - `rundll32.exe`).

Some recommendations here but open to other ways: https://github.com/microsoft/WSL/issues/423#issuecomment-221627364

**Pull request merged**